### ISSUE TITLE
Type narrowing behaves oddly in presence of type.__eq__

### ISSUE THREAD

**Issue by mniip** [microsoft/pyright#4419]
**Describe the bug**
In some situations an `x: Union[X, Y, ...]` that has been previously narrowed to `Y` incorrectly unnarrows back to `x: Union[X, Y]`. This happens seemingly when we branch on `something == X`.

**To Reproduce**
Run pyright on the code from the Code section.
```
  a.py:12:18 - error: Cannot access member "p" for type "X"
    Member "p" is unknown (reportGeneralTypeIssues)
  a.py:11:25 - information: Type of "x" is "Y"
```

**Expected behavior**
No error.

**Screenshots or Code**
```py
from typing import Union

class X:
    pass
class Y:
    p: type

def f(xs: list[Union[X, Y]]) -> None:
    for x in xs:
        if not isinstance(x, X):
            reveal_type(x)
            if x.p == X:
                pass
```

**VS Code extension or command-line**
Commandline pyright 1.1.287

**Additional context**
Works on pyright 1.1.285

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.288, which I just published. It will also be included in a future release of pylance.
